File: leetcode/cpp/962.maximum-width-ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        vector<int> stack;
        for (int i=0; i < nums.size(); i++) {
            if (stack.empty() || nums[stack.back()] >= nums[i]) {
                stack.push_back(i);
            }
        }
        int ans = 0;
        for (int i = nums.size() - 1; i > ans; --i) {
            while (!stack.empty() && nums[i] >= nums[stack.back()])
                ans = max(ans, i - stack.back()), stack.pop_back();
        }
        return ans;
    }
};

File: leetcode/python/962.maximum-width-ramp.py
class Solution:
    def maxWidthRamp(self, nums: List[int]) -> int:
        stack = []
        ans = 0
        for i, n in enumerate(nums):
            if not stack or nums[stack[-1]] >= n:
                stack.append(i)
        
        for i, n in reversed(list(enumerate(nums))):
            while stack and n >= nums[stack[-1]]:
                ans = max(ans, i - stack.pop())

        return ans

